Add some explanations on `NonInterpPrimitive` class (#6851)
PennyLaneAI/pennylane: pennylane/capture/explanations.md

> [This](This) documentation explains the principles behind `qml.capture.CaptureMeta` and higher order primitives.

Author: Whoops, what hotkey did I strike here :o

def fun(x):
    return (x**2, 4 * x - 3, x**23)
```

Now suppose we want to turn this into a primitive. We could just promote it to a standard
`jax.core.Primitive` as

```python
import jax

fd_prim = jax.core.Primitive("finite_diff")
fd_prim.multiple_results = True
fd_prim.def_impl(finite_diff_impl)

def finite_diff(x, fun, delta=1e-5):
    return fd_prim.bind(x, fun, delta)
```

This allows us to use the forward pass as usual (to compute the first-order derivative):

```pycon
>>> finite_diff(1., fun, delta=1e-6)
(2.000000000002, 3.999999999892978, 23.000000001216492)
```

Now if we want to make this primitive differentiable (with automatic
differentiation/backprop, not by using a higher order finite difference scheme),
we need to specify a JVP rule. (Note that there are multiple rather simple fixes for this example
that we could use to implement a finite difference scheme that is readily differentiable. This is
somewhat besides the point, because we did not identify a possibility to use any of those
alternatives in the PennyLane code).

However, the finite difference rule is just a standard
algebraic function making use of calls to `fun` and some elementary operations, so ideally

> we would like to just use the chain rule as it is known to the AD engine. A JVP rule would

Reviewer: I never heard the expression 'AD engine'. Although I have an intuitive understanding of its meaning, could you please add a clarification? Thank you!

Author: I changed it to the more standard "automatic differentiation framework" :)

> we would like to just use the chain rule as it is known to the AD engine. A JVP rule would

Author: ```suggestion
we would like to just use the chain rule as it is known to the automatic differentiation framework. A JVP rule would
```

> Now if we want to make this primitive differentiable (with automatic
> differentiation/backprop, not by using a higher order finite difference scheme),
> we need to specify a JVP rule. (Note that there are multiple rather simple fixes for this example
> that we could use to implement a finite difference scheme that is readily differentiable. This is
> somewhat besides the point, because we did not identify a possibility to use any of those
> alternatives in the PennyLane code).
> 
> However, the finite difference rule is just a standard
> algebraic function making use of calls to `fun` and some elementary operations, so ideally
> we would like to just use the chain rule as it is known to the AD engine. A JVP rule would
> then just manually re-implement this chain rule, which we'd rather not do.
> 
> Instead, we define a non-interpreted type of primitives, and create such a primitive
> for our finite difference method. We also create the usual method that binds the
> primitive to inputs.

Reviewer: ```suggestion
Now if we want to make this primitive differentiable (with automatic
differentiation/backprop, not by using a higher-order finite difference scheme),
we need to specify a JVP rule. (Note that there are multiple rather simple fixes for this example
that we could use to implement a finite difference scheme that is readily differentiable. This is
somewhat beside the point because we did not identify the possibility of using any of those
alternatives in the PennyLane code).

However, the finite difference rule is just a standard
algebraic function making use of calls to `fun` and some elementary operations, so ideally
we would like to just use the chain rule as it is known to the AD engine. A JVP rule would
then just manually re-implement this chain rule, which we'd rather not do.

Instead, we define a non-interpreted type of primitive and create such a primitive
for our finite difference method. We also create the usual method that binds the
primitive to inputs.
```

```python
class NonInterpPrimitive(jax.core.Primitive):
    """A subclass to JAX's Primitive that works like a Python function
    when evaluating JVPTracers."""

    def bind_with_trace(self, trace, args, params):
        """Bind the ``NonInterpPrimitive`` with a trace.
        If the trace is a ``JVPTrace``, it falls back to a standard Python function call.
        Otherwise, the bind call of JAX's standard Primitive is used."""
        if isinstance(trace, jax.interpreters.ad.JVPTrace):

>             return self.impl(*args, **params)

Reviewer 2: What if we have nested tracers? `jax.grad(jax.jit(f))`? Or something like that. 

Though I guess we can just proceed with this till we find an issue.

Author: Still seems to work :)

        return super().bind_with_trace(trace, args, params)

fd_prim_2 = NonInterpPrimitive("finite_diff_2")
fd_prim_2.multiple_results = True
fd_prim_2.def_impl(finite_diff_impl) # This also defines the behaviour with a JVP tracer

def finite_diff_2(x, fun, delta=1e-5):
    return fd_prim_2.bind(x, fun, delta)
```

Now we can use the primitive in a differentiable workflow, without defining a JVP rule
that just repeats the chain rule:

```pycon
>>> # Forward execution of finite_diff_2 (-> first-order derivative)

> >>> finite_diff_2(fun, delta=1e-6)(1.)

Reviewer: ```suggestion
>>> finite_diff_2(1., fun, delta=1e-6)
```

(2.000000000002, 3.999999999892978, 23.000000001216492)
>>> # Differentiation of finite_diff_2 (-> second-order derivative)
>>> jax.jacobian(finite_diff_2)(1., fun, delta=1e-6)
(Array(1.9375, dtype=float32, weak_type=True), Array(0., dtype=float32, weak_type=True), Array(498., dtype=float32, weak_type=True))
```

> In addition to the differentiation primitives for `qml.jacobian` and `qml.grad`, quantum operators
> have non-interpreted primitives as well. This is because their differentiation is performed
> by the surrounding QNode primitive rather than through the standard chain rule that acts
> "locally" (in the circuit). In short, we only want gates to store their tracers (which will help
> determining differentiability of gate arguments, for example), but not to do anything with them.

Reviewer: ```suggestion
In addition to the differentiation primitives for `qml.jacobian` and `qml.grad`, quantum operators
have non-interpreted primitives as well. This is because their differentiation is performed
by the surrounding QNode primitive rather than through the standard chain rule that acts
"locally" (in the circuit). In short, we only want gates to store their tracers (which will help
determine the differentiability of gate arguments, for example), but not to do anything with them.
```

we would like to just use the chain rule as it is known to the AD engine. A JVP rule would
then just manually re-implement this chain rule, which we'd rather not do.

Instead, we define a non-interpreted type of primitives, and create such a primitive
for our finite difference method. We also create the usual method that binds the
primitive to inputs.

```python
class NonInterpPrimitive(jax.core.Primitive):
    """A subclass to JAX's Primitive that works like a Python function
    when evaluating JVPTracers."""

    def bind_with_trace(self, trace, args, params):
        """Bind the ``NonInterpPrimitive`` with a trace.
        If the trace is a ``JVPTrace``, it falls back to a standard Python function call.
        Otherwise, the bind call of JAX's standard Primitive is used."""
        if isinstance(trace, jax.interpreters.ad.JVPTrace):
            return self.impl(*args, **params)
        return super().bind_with_trace(trace, args, params)

fd_prim_2 = NonInterpPrimitive("finite_diff_2")
fd_prim_2.multiple_results = True
fd_prim_2.def_impl(finite_diff_impl) # This also defines the behaviour with a JVP tracer

def finite_diff_2(fun, delta=1e-5):

>     return fd_prim_2.bind(fun, delta)

Author: Thanks!

> def finite_diff_2(fun, delta=1e-5):
>     return fd_prim_2.bind(fun, delta)

Author: ```suggestion
def finite_diff_2(x, fun, delta=1e-5):
    return fd_prim_2.bind(x, fun, delta)
```

```

Now we can use the primitive in a differentiable workflow, without defining a JVP rule
that just repeats the chain rule:

```pycon
>>> # Forward execution of finite_diff_2 (-> first-order derivative)
>>> finite_diff_2(fun, delta=1e-6)(1.)
(2.000000000002, 3.999999999892978, 23.000000001216492)
>>> # Differentiation of finite_diff_2 (-> second-order derivative)
>>> jax.jacobian(finite_diff_2)(1., fun, delta=1e-6)
(Array(1.9375, dtype=float32, weak_type=True), Array(0., dtype=float32, weak_type=True), Array(498., dtype=float32, weak_type=True))
```

In addition to the differentiation primitives for `qml.jacobian` and `qml.grad`, quantum operators
have non-interpreted primitives as well. This is because their differentiation is performed
by the surrounding QNode primitive rather than through the standard chain rule that acts
"locally" (in the circuit). In short, we only want gates to store their tracers (which will help

> determining differentiability of gate arguments, for example), but not to do anything with them.

Author: To be honest, I am not entirely confident in deriving such a rule myself.
It sounds reasonable but I have not thought much further than about the original `grad/jacobian` related scope :sweat_smile:

> determining differentiability of gate arguments, for example), but not to do anything with them.

Reviewer 3: Fair enough. It seems fine to me for a couple of reasons:
* Operators and measurements don't really have meaningful use outside a `QNode`, which has a JVP rule, so they can be "safely" assumed to be `NonInterpPrimitive`s.
* HOPs can be `NonInterpPrimitive`s for basically the same reason. Any quantum instructions present inside these primitives are only meaningful if they're inside a `QNode`, and classical instructions will be interpreted automatically on the forward pass.

> determining differentiability of gate arguments, for example), but not to do anything with them.

Author: Yeah, that does sound reasonable :)

> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

Reviewer 2: ```suggestion
```

> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

Author: ```suggestion
```

> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

Author: ```suggestion
```
